includes/abilities-api/class-wp-ability.php

```diff
@@ -336,14 +336,38 @@ public function execute( array $input = array() ) {
 			);
 		}
 
+		/**
+		 * Fires before an ability gets executed.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param string              $ability_name The name of the ability.
+		 * @param array<string,mixed> $input        The input data for the ability.
+		 */
+		do_action( 'before_execute_ability', $this->name, $input );
+
 		$result = $this->do_execute( $input );
 		if ( is_wp_error( $result ) ) {
 			return $result;
 		}
 
 		$is_valid = $this->validate_output( $result );
+		if ( is_wp_error( $is_valid ) ) {
+			return $is_valid;
+		}
 
-		return is_wp_error( $is_valid ) ? $is_valid : $result;
+		/**
+		 * Fires immediately after an ability finished executing.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param string              $ability_name The name of the ability.
+		 * @param array<string,mixed> $input        The input data for the ability.
+		 * @param mixed               $result       The result of the ability execution.
+		 */
+		do_action( 'executed_ability', $this->name, $input, $result );
```

Comment:
Worried this name is a footgun, since that `*d_*` is hard to spot and WP core sometimes uses present-tense, so it would be easy to assume `execute_ability` (no `d`) was correct.

```suggestion
		do_action( 'after_execute_ability', $this->name, $input, $result );
```

Reply:
I don't mind using `after_` & `before_` pair here. There is no strict pattern established in WordPress core.